pkg/sql/compile/cnserviceDriver.go

```diff
@@ -51,104 +51,182 @@ import (
 	if err != nil {
 		return nil, err
 	}
-	// refactor the last operator connect to output
-	refactorScope(ctx, s, cs)
+	refactorScope(c, c.ctx, s)
 
-	err = s.Run(c)
+	err = s.ParallelRun(c)
 	if err != nil {
 		return nil, err
 	}
-	// get analyse related information
-	if query, ok := s.Plan.Plan.(*plan.Plan_Query); ok {
-		nodes := query.Query.GetNodes()
-		anas := &pipeline.AnalysisList{}
-		anas.List = make([]*plan.AnalyzeInfo, len(nodes))
-		for i := range anas.List {
-			anas.List[i] = nodes[i].AnalyzeInfo
-		}
-		anaData, err = anas.Marshal()
-		if err != nil {
-			return nil, err
-		}
-		return anaData, nil
+	// encode analysis info and return.
+	anas := &pipeline.AnalysisList{
+		List: make([]*plan.AnalyzeInfo, len(c.proc.AnalInfos)),
+	}
+	for i := range anas.List {
+		anas.List[i] = convertToPlanAnalyzeInfo(c.proc.AnalInfos[i])
 	}
-	return nil, nil
+	return anas.Marshal()
 }
 
-// remoteRun sends a scope to a remote node for execution, and wait to receive the back message.
-// the back message is always *pipeline.Message but has three cases.
-// 1. ErrMessage
-// 2. End Message with the result of analysis
-// 3. Batch Message
+// remoteRun sends a scope to remote node for execution, and wait to receive the back message.
+// the back message is always *pipeline.Message but contains three cases.
+// 1. Message with error information
+// 2. Message with end flag and the result of analysis
+// 3. Batch Message with batch data
 func (s *Scope) remoteRun(c *Compile) error {
 	// encode the scope
+	// the last instruction of remote-run scope must be `connect`, it doesn't need serialization work.
+	// just ignore this instruction when doing serialization work and recover at the end in order to receive the returned batch.
+	n := len(s.Instructions) - 1
+	con := s.Instructions[n]
+	s.Instructions = s.Instructions[:n]
 	sData, errEncode := encodeScope(s)
 	if errEncode != nil {
 		return errEncode
 	}
+	s.Instructions = append(s.Instructions, con)
+
+	// encode the process related information
+	pData, errEncodeProc := encodeProcessInfo(c.proc)
+	if errEncodeProc != nil {
+		return errEncodeProc
+	}
+
+	// get the stream-sender
+	streamSender, errStream := cnclient.GetStreamSender(s.NodeInfo.Addr)
+	if errStream != nil {
+		return errStream
+	}
+	defer func(streamSender morpc.Stream) {
+		_ = streamSender.Close()
+	}(streamSender)
 
 	// send encoded message
-	message := &pipeline.Message{Data: sData}
-	r, errSend := cnclient.Client.Send(c.ctx, s.NodeInfo.Addr, message)
+	// mo-rpc send message requires that context should have its own timeout.
+	// TODO: get dead time from config file may suitable.
+	if _, ok := c.ctx.Deadline(); !ok {
+		var cancel context.CancelFunc
+		c.ctx, cancel = context.WithTimeout(c.ctx, time.Second*10000)
+		_ = cancel
+	}
+
+	message := cnclient.AcquireMessage()
+	{
+		message.Id = streamSender.ID()
+		message.Data = sData
+		message.ProcInfoData = pData
+	}
+	defer cnclient.ReleaseMessage(message)
```

Comment:
cnclient.ReleaseMessage called here.
after the client get all data from remote-node. we will release the Message @zhangxu19830126